engine/lib/membrane_rtc_engine/engine.ex

```diff
@@ -1218,11 +1239,21 @@ defmodule Membrane.RTC.Engine do
   #   owned by one of the Endpoints in the list.
   #
 
-  defp validate_subscription(subscription, state) do
+  defp validate_subscription(subscription, ctx, state) do
     # checks whether subscription is correct
     track = get_track(subscription.track_id, state.endpoints)
 
-    if track, do: :ok, else: {:error, :invalid_track_id}
+    endpoint_id = {:endpoint, subscription.endpoint_id}
+
+    endpoint = find_child(ctx, pattern: ^endpoint_id)
+    endpoint_in_state? = Map.has_key?(state.endpoints, subscription.endpoint_id)
+
+    cond do
+      is_nil(track) -> {:error, :invalid_track_id}
+      not endpoint_in_state? and is_nil(endpoint) -> {:error, :endpoint_not_exist}
+      not endpoint_in_state? -> {:error, :endpoint_removed}
```

Comment:
Does it mean that we never clear the state from added endpoints? Determining whether something was removed or never existed is not a trivial task. That's why we always return `:invalid_track_id`. Right now, the API is not fully symmetric. Can't we just return `:invalid_endpoint_id` and let user decide what to do?

Reply:
Nope, we cleaned the state from added endpoints. The problem with returning `:invalid_endpoint_id` is that the endpoint lacks information to properly handle each of these cases because IMO `:endpoint_removed` and `:endpoint_not_exist` are totally different cases and should be handled differently. `endpoint_not_exist` means that endpoint is not in the engine which means that this is a critical bug, because e.g: we pass the wrong pid as `rtc_engine`. `endpoint_removed` on the other hand is simply handling a race condition which can happen that endpoint sends subscribe request after it was removed from the engine and in that case, endpoint should ignore that it doesn't subscribe on track.